wpimath/src/main/java/edu/wpi/first/math/geometry/Quaternion.java

```diff
@@ -126,14 +187,110 @@ public Quaternion inverse() {
+   *
+   * @param t the power to raise this quaternion to.
+   * @return The quaternion power
+   */
+  public Quaternion pow(double t) {
+    // q^t = e^(ln(q^t)) = e^(t * ln(q))
+    return this.log().times(t).exp();
+  }
+
+  /**
+   * Matrix exponential of a quaternion.
+   *
+   * @param adjustment the "Twist" that will be applied to this quaternion.
+   * @return The quaternion product of exp(adjustment) * this
+   */
+  public Quaternion exp(Quaternion adjustment) {
+    return adjustment.exp().times(this);
+  }
+
+  /**
+   * Matrix exponential of a quaternion.
+   *
+   * <p>source: https://en.wikipedia.org/wiki/Quaternion#Exponential,_logarithm,_and_power_functions
+   *
+   * @return The Matrix exponential of this quaternion.
+   */
+  public Quaternion exp() {
+    // q = s(scalar) + v(vector)
+    // exp(s)
+    var scalar = Math.exp(m_w);
+
+    // ||v||
+    var axial_magnitude = Math.sqrt(getX() * getX() + getY() * getY() + getZ() * getZ());
+    // cos(||v||)
+    var cosine = Math.cos(axial_magnitude);
+
+    double axial_scalar;
+
+    if (axial_magnitude < 1e-9) {
+      // Taylor series of sin(x)/x near x=0: 1 − x²/6 + x⁴/120 + O(n⁶)
+      var axial_magnitude_sq = axial_magnitude * axial_magnitude;
+      var axial_magnitude_sq_sq = axial_magnitude_sq * axial_magnitude_sq;
+      axial_scalar = 1 - axial_magnitude_sq / 6 + axial_magnitude_sq_sq / 120;
+    } else {
+      axial_scalar = Math.sin(axial_magnitude) / axial_magnitude;
+    }
+
+    // exp(s) * (cos(||v||) + v * sin(||v||) / ||v||)
+    return new Quaternion(
+        cosine * scalar,
+        getX() * axial_scalar * scalar,
+        getY() * axial_scalar * scalar,
+        getZ() * axial_scalar * scalar);
+  }
+
+  /**
+   * Log operator of a quaternion.
+   *
+   * @param end The quaternion to map this quaternion onto.
+   * @return The "Twist" that maps this quaternion to the argument.
+   */
+  public Quaternion log(Quaternion end) {
+    return end.times(this.inverse()).log();
+  }
+
+  /**
+   * The Log operator of a general quaternion.
+   *
+   * <p>source: https://github.com/moble/quaternionic/blob/main/quaternionic/algebra.py#L233
+   *
+   * <p>For unit quaternions, this is equivalent to @see Quaternion#toRotationVector().
+   *
+   * @return The logarithm of this quaternion.
+   */
+  public Quaternion log() {
+    var vnorm = Math.sqrt(getX() * getX() + getY() * getY() + getZ() * getZ());
+    if (vnorm <= 1e-9 * Math.abs(getW())) {
+      if (getW() < 0) {
+        if (Math.abs(getW() + 1) > 1e-9) {
+          return new Quaternion(Math.log(-getW()), Math.PI, 0, 0);
+        } else {
+          return new Quaternion(0, Math.PI, 0, 0);
+        }
+      } else {
+        return new Quaternion(Math.log(getW()), 0, 0, 0);
+      }
+    } else {
+      var v = Math.atan2(vnorm, getW());
+      var f = v / vnorm;
+      return new Quaternion(
+          Math.log(getW() * getW() + vnorm * vnorm) / 2.0, f * getX(), f * getY(), f * getZ());
```

Comment:
Could we please get an explanation of the math somewhere (here or `wpimath/algorithms.md` if it gets too long)? The linked source code doesn't have any.

Here's a brief sketch of why the math \*should\* work, though this would need to be reworked before actually putting it in the code (e.g., use Unicode math symbols, derive the formula rather than reasoning why it works, etc.).
```
Define `||q_v||` to be `sqrt(q_x^2 + q_y^2 + q_z^2)`.

Take as given that `exp(q) = exp(q_w) (cos(||q_v||) + sin(||q_v||) (q_x i + q_y j + q_z k) / ||q_v||)`

  exp(ln(q_w^2 + ||q_v||^2) / 2 + atan2(||q_v||, q_w) (q_x i + q_y j + q_z k) / ||q_v||))
    = exp(ln(q_w^2 + ||q_v||^2) / 2) (cos(atan2(||q_v||, q_w)) + sin(atan2(||q_v||, q_w)) (q_x i + q_y j + q_z k ) / ||q_v||)
    = exp(ln(sqrt(q_w^2 + ||q_v||^2))) (q_w / (||q_v||^2 + q_w^2) + (||q_v|| / (||q_v||^2 + q_w^2)) (q_x i + q_y j + q_z k) / ||q_v||)
    = sqrt(q_w^2 + ||q_v||^2) (q_w / (q_w^2 + ||q_v||^2) + (q_x i + q_y j + q_z k) / (q_w^2 + ||q_v||^2))
    = q_w + q_x i + q_y j + q_z k
```

Reply:
Done - details have been added for the general quaternion log and exp, as well as toRotationVector and fromRotationVector to `wpimath/algorithms.md`